# HG changeset patch
# Parent  97f7fd10087f08547f48c20644461f2356e7399b
Handle bulbs with bugged tags bitfields

I'm getting that on the LIFX White 800 for tag and untag I'm afraid
something is weird with their firmare.

diff --git a/core/proto.c b/core/proto.c
--- a/core/proto.c
+++ b/core/proto.c
@@ -209,10 +209,20 @@
         bool comma = false;
         int tag_id;
         LGTD_LIFX_WIRE_FOREACH_TAG_ID(tag_id, bulb->state.tags) {
-            lgtd_client_write_string(client, comma ? ",\"" : "\"");
-            lgtd_client_write_string(client, bulb->gw->tags[tag_id]->label);
-            lgtd_client_write_string(client, "\"");
-            comma = true;
+            if (LGTD_LIFX_WIRE_TAG_ID_TO_VALUE(tag_id) & bulb->gw->tag_ids) {
+                lgtd_client_write_string(client, comma ? ",\"" : "\"");
+                lgtd_client_write_string(client, bulb->gw->tags[tag_id]->label);
+                lgtd_client_write_string(client, "\"");
+                comma = true;
+            } else {
+                lgtd_warnx(
+                    "tag_id %d on bulb %.*s (%s) doesn't "
+                    "exist on gw [%s]:%hu (site %s)",
+                    tag_id, (int)sizeof(bulb->state.label), bulb->state.label,
+                    lgtd_addrtoa(bulb->addr), bulb->gw->ip_addr, bulb->gw->port,
+                    lgtd_addrtoa(bulb->gw->site.as_array)
+                );
+            }
         }
 
         lgtd_client_write_string(
